Sources/YUL/AST.swift

```diff
@@ -0,0 +1,342 @@
+  case s64
+  case u128
+  case s128
+  case u256
+  case s256
+  case any
+
+  public var description: String {
+    return self.rawValue
+  }
+}
+
+public typealias TypedIdentifierList = [(Identifier, Type)]
+
+private func render(typedIdentifierList: TypedIdentifierList) -> String {
+  return typedIdentifierList.map({ (ident, type) in
+    switch type {
+    case .any:
+      return "\(ident)"
+    default:
+      return "\(ident): \(type)"
+    }
+  }).joined(separator: ", ")
+}
+
+public struct FunctionDefinition: CustomStringConvertible {
+  public let identifier: Identifier
+  public let arguments: TypedIdentifierList
+  public let returns: TypedIdentifierList
+  public let body: Block
+
+  public init(identifier: Identifier,
+              arguments: TypedIdentifierList,
+              returns: TypedIdentifierList = [],
+              body: Block) {
+    self.identifier = identifier
+    self.arguments = arguments
+    self.returns = returns
+    self.body = body
+  }
+
+  public var description: String {
+    let args = render(typedIdentifierList: self.arguments)
+
+    var ret = ""
+    if !self.returns.isEmpty {
+      let retargs = render(typedIdentifierList: self.returns)
+      ret = "-> \(retargs)"
+    }
+
+    return "\(self.identifier)(\(args)) \(ret) \(self.body)"
+  }
+}
+
+public struct VariableDeclaration: CustomStringConvertible {
+  public let declarations: TypedIdentifierList
+  public let expression: Expression?
+
+  public init(_ declarations: TypedIdentifierList, _ expression: Expression? = nil) {
+    self.declarations = declarations
+    self.expression = expression
+  }
+
+  public var catchableSuccesses: [Expression] {
+    return expression?.catchableSuccesses ?? []
+  }
+
+  public var description: String {
+    let decls = render(typedIdentifierList: self.declarations)
+    if self.expression == nil {
+      return "let \(decls)"
+    }
+    return "let \(decls) := \(self.expression!.description)"
+  }
+}
+
+public typealias IdentifierList = [Identifier]
+
+public struct Assignment: CustomStringConvertible {
+  public let identifiers: IdentifierList
+  public let expression: Expression
+
+  public init(_ identifiers: IdentifierList, _ rhs: Expression) {
+    self.identifiers = identifiers
+    self.expression = rhs
+  }
+
+  public var catchableSuccesses: [Expression] {
+    return expression.catchableSuccesses
+  }
+
+  public var description: String {
+    let lhs = self.identifiers.joined(separator: ", ")
+    return "\(lhs) := \(self.expression)"
+  }
+}
+
+public indirect enum Expression: CustomStringConvertible {
+  case functionCall(FunctionCall)
+  case identifier(Identifier)
+  case literal(Literal)
+  case catchable(value: Expression, success: Expression)
+
+  // TODO: these three should really be statements
+  case variableDeclaration(VariableDeclaration)
+  case assignment(Assignment)
+  case noop
+
+  case inline(String)
+
+  public var catchableSuccesses: [Expression] {
+    switch self {
```

Comment:
Can you avoid this & other switches with a catchableSuccesses protocol?

Reply:
I can add the protocol, but the switches are still necessary, since they are different for each `struct`.

Reply:
ok yeah a protocol would be nice and then we can think about how to reengineer this for less duplication